src/core/layout/qgslayoutitemmap.cpp

```diff
@@ -34,7 +80,1631 @@ QString QgsLayoutItemMap::stringType() const
+
+  //ensure polygon is closed by readding first point
+  poly << QPointF( poly.at( 0 ) );
+}
+
+void QgsLayoutItemMap::transformShift( double &xShift, double &yShift ) const
+{
+  double mmToMapUnits = 1.0 / mapUnitsToLayoutUnits();
+  double dxScaled = xShift * mmToMapUnits;
+  double dyScaled = - yShift * mmToMapUnits;
+
+  QgsLayoutUtils::rotate( mEvaluatedMapRotation, dxScaled, dyScaled );
+
+  xShift = dxScaled;
+  yShift = dyScaled;
+}
+
+void QgsLayoutItemMap::drawAnnotations( QPainter *painter )
+{
+  if ( !mLayout || !mLayout->project() || !mDrawAnnotations )
+  {
+    return;
+  }
+
+  const QList< QgsAnnotation * > annotations = mLayout->project()->annotationManager()->annotations();
+  if ( annotations.isEmpty() )
+    return;
+
+  QgsRenderContext rc = QgsLayoutUtils::createRenderContextForMap( this, painter );
+  rc.setForceVectorOutput( true );
+  rc.setExpressionContext( createExpressionContext() );
+  QList< QgsMapLayer * > layers = layersToRender( &rc.expressionContext() );
+
+  for ( QgsAnnotation *annotation : annotations )
+  {
+    if ( !annotation || !annotation->isVisible() )
+    {
+      continue;
+    }
+    if ( annotation->mapLayer() && !layers.contains( annotation->mapLayer() ) )
+      continue;
+
+    drawAnnotation( annotation, rc );
+  }
+}
+
+void QgsLayoutItemMap::drawAnnotation( const QgsAnnotation *annotation, QgsRenderContext &context )
+{
+  if ( !annotation || !annotation->isVisible() || !context.painter() || !context.painter()->device() )
+  {
+    return;
+  }
+
+  context.painter()->save();
+  context.painter()->setRenderHint( QPainter::Antialiasing, context.flags() & QgsRenderContext::Antialiasing );
+
+  double itemX, itemY;
+  if ( annotation->hasFixedMapPosition() )
+  {
+    QPointF mapPos = layoutMapPosForItem( annotation );
+    itemX = mapPos.x();
+    itemY = mapPos.y();
+  }
+  else
+  {
+    itemX = annotation->relativePosition().x() * rect().width();
+    itemY = annotation->relativePosition().y() * rect().height();
+  }
+  context.painter()->translate( itemX, itemY );
+
+  //setup painter scaling to dots so that symbology is drawn to scale
+  double dotsPerMM = context.painter()->device()->logicalDpiX() / 25.4;
+  context.painter()->scale( 1 / dotsPerMM, 1 / dotsPerMM ); // scale painter from mm to dots
+
+  annotation->render( context );
+  context.painter()->restore();
+}
+
+QPointF QgsLayoutItemMap::layoutMapPosForItem( const QgsAnnotation *annotation ) const
+{
+  if ( !annotation )
+    return QPointF( 0, 0 );
+
+  double mapX = 0.0;
+  double mapY = 0.0;
+
+  mapX = annotation->mapPosition().x();
+  mapY = annotation->mapPosition().y();
+  QgsCoordinateReferenceSystem annotationCrs = annotation->mapPositionCrs();
+
+  if ( annotationCrs != crs() )
+  {
+    //need to reproject
+    QgsCoordinateTransform t( annotationCrs, crs() );
+    double z = 0.0;
+    try
+    {
+      t.transformInPlace( mapX, mapY, z );
+    }
+    catch ( const QgsCsException & )
+    {
```

Comment:
There's no error message raised from this exception. Is it wanted?

Reply:
I don't think so - it will log a message in debug mode, but really the exception is only caught here to avoid a crash.